components/campaign/CampaignChaosBagView/index.tsx

```diff
@@ -0,0 +1,382 @@
+            id: 'back',
+            color: COLORS.navButton,
+            testID: t`Cancel`,
+          } : {
+            icon: iconsMap['arrow-left'],
+            id: 'androidBack',
+            color: COLORS.navButton,
+            testID: t`Back`,
+          },
+        ],
+        rightButtons: [{
+          systemItem: 'save',
+          text: t`Edit`,
+          id: 'edit',
+          showAsAction: 'ifRoom',
+          color: COLORS.navButton,
+          testID: t`Edit`,
+        }],
+      },
+    };
+  }
+
+  _navEventListener?: EventSubscription;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      chaosBagResults: {
+        currentToken: undefined,
+        drawnTokens: [],
+        total: 0,
+      },
+      visible: true,
+    };
+
+    this._navEventListener = Navigation.events().bindComponent(this);
+  }
+
+  componentDidMount() {
+    BackHandler.addEventListener('hardwareBackPress', this._handleBackPress);
+  }
+
+  componentWillUnmount() {
+    BackHandler.removeEventListener('hardwareBackPress', this._handleBackPress);
+    this._navEventListener && this._navEventListener.remove();
+  }
+
+  componentDidAppear() {
+    this.setState({
+      visible: true,
+    });
+  }
+
+  componentDidDisappear() {
+    this.setState({
+      visible: false,
+    });
+  }
+
+  navigationButtonPressed({ buttonId }: { buttonId: string }) {
+    if (buttonId === 'back' || buttonId === 'androidBack') {
+      this._handleBackPress();
+    } else if (buttonId === 'edit') {
+      this._showChaosBagDialog();
+    }
+  }
+
+  _handleDrawTokenPressed = () => {
+    this._drawToken();
+  }
+
+  _handleSealTokenPressed = () => {
+
+  }
+
+  _handleBackPress = () => {
+    const {
+      componentId,
+    } = this.props;
+    const {
+      visible,
+    } = this.state;
+    if (!visible) {
+      return false;
+    }
+    Navigation.pop(componentId);
+    return true;
+  }
+
+  _showChaosBagDialog = () => {
+    const {
+      componentId,
+      updateChaosBag,
+      chaosBag,
+    } = this.props;
+
+    Navigation.push<EditChaosBagProps>(componentId, {
+      component: {
+        name: 'Dialog.EditChaosBag',
+        passProps: {
+          chaosBag,
+          updateChaosBag: updateChaosBag,
+          trackDeltas: true,
+        },
+        options: {
+          topBar: {
+            title: {
+              text: t`Chaos Bag`,
+            },
+            backButton: {
+              title: t`Cancel`,
+            },
+          },
+        },
+      },
+    });
+  }
+
+  _getRandomNumber = (min: number, max: number) => {
+    return Math.random() * (max - min) + min;
+  }
+
+  _getRandomChaosToken = (chaosBag: ChaosBag) => {
+    const list = keys(chaosBag);
+    const weight = values(chaosBag);
+
+    const totalWeight = weight.reduce((prev, cur) => {
+      return (prev && cur) ? (prev + cur) : 0;
+    });
+
+    if (totalWeight) {
+      const random_num = this._getRandomNumber(0, totalWeight);
```

Comment:
Given our set is always pretty small (14-20 tokens), it might be simpler to just make an array of all available tokens, `shuffle` them (using lodash) and take the first?  

If they draw another token, just repeat the above (with one less available now)?

I'm sure this method would work as well, it's just a bit harder to reason about the corner cases in it, and random things can be hard to test.